File: code/mofa_export.py
#!/usr/bin/env python

import mofax as mofa
import muon as mu
import pandas as pd
import os
import argparse

# --- Parse arguments ---
parser = argparse.ArgumentParser(description="Export MOFA+ factors and loadings")
parser.add_argument('--model', default='results/mofa/t1.hdf5')
parser.add_argument('--mdata', default='results/mofa/mdata.h5mu')
parser.add_argument('--outdir', default='results/mofa')
args = parser.parse_args()

os.makedirs(args.outdir, exist_ok=True)

# --- Load model and mdata ---
model = mofa.mofa_model(args.model)
mdata = mu.read(args.mdata)

# --- Export factors ---
factors_arr = model.get_factors()
sample_index = model.get_samples()

# Check shape match
if len(sample_index) != factors_arr.shape[0]:
    print(f"Error: Sample count mismatch. {len(sample_index)} sample names vs {factors_arr.shape[0]} factor rows.")
    exit()

# Extract just the 'sample' column if sample_index is a DataFrame
if isinstance(sample_index, pd.DataFrame):
    sample_index = sample_index.iloc[:, 1]  # or sample_index["sample"]

factors_df = pd.DataFrame(
       factors_arr,
       index=sample_index,
       columns=[f"Factor_{i+1}" for i in range(model.nfactors)]
       )

factors_df.to_csv(os.path.join(args.outdir, 'sample_factors.tsv'), sep='\t')
print("Exported sample factors.")

# --- Export loadings ---
weights = model.get_weights()
if not isinstance(weights, dict):
    weights = {model.get_views()[0]: weights}

for view_name, weights_arr in weights.items():
    try:
        features = mdata[view_name].var_names
    except KeyError:
        print(f"Skipping {view_name}, features not found.")
        continue

    if weights_arr.shape[0] != len(features):
        print(f"Skipping {view_name}, shape mismatch.")
        print(f"  weights_arr.shape[0] = {weights_arr.shape[0]}, len(features) = {len(features)}")
        continue

    df = pd.DataFrame(
        weights_arr,
        index=features,
        columns=[f"Factor_{i+1}" for i in range(model.nfactors)]
    )
    df.to_csv(os.path.join(args.outdir, f'loadings_{view_name}.tsv'), sep='\t')
    print(f"Exported loadings for {view_name}")

File: code/mofa_export.R
#!/usr/bin/env Rscript
library(MOFA2)
library(optparse)

option_list <- list(
  make_option(c("--model"), type="character", default=NULL),
  make_option(c("--outdir"), type="character", default="extract")
)
opt <- parse_args(OptionParser(option_list=option_list))

dir.create(opt$outdir, showWarnings=FALSE, recursive=TRUE)

# Load model
model <- load_model(opt$model)

# Save factors
factors <- get_factors(model)

# Fix row names and column names
factors_df <- as.data.frame(factors)
factors_df$sampleID <- rownames(factors_df)
rownames(factors_df) <- NULL
colnames(factors_df) <- sub("^group1\\.", "", colnames(factors_df))

# Move sampleID to first column
factors_df <- factors_df[, c("sampleID", setdiff(names(factors_df), "sampleID"))]

write.table(factors_df,
            file=file.path(opt$outdir, "sample_factors.tsv"),
            sep="\t", quote=FALSE, row.names=FALSE)

# Save weights per view with Feature as first column
view_names <- names(model@data)  # get the views in the MOFA2 model
for (v in view_names) {
  w <- get_weights(model, views=v)
  df <- as.data.frame(w[[1]])
  df <- cbind(Feature=rownames(df), df)  # Feature as first column
  write.table(df,
              file=file.path(opt$outdir, paste0("loadings_", v, ".tsv")),
              sep="\t", quote=FALSE, row.names=FALSE)
}

